sale_order_product_assortment/models/sale_order.py

```diff
@@ -37,3 +38,25 @@ def _compute_product_assortment_ids(self):
                     )
             if self.allowed_product_ids:
                 self.has_allowed_products = True
+
+    @api.constrains("allowed_product_ids", "order_line.product_id")
```

Comment:
```suggestion
    @api.onchange("partner_id")
```
I would prefer this check to be carried out when changing the partner, as this is really the source of the problem.